### DISCUSSION TITLE
Define the event and time in cvodes

### DISCUSSION THREAD

**Discussion by PrakitrSrisuma** [casadi/casadi#3105] [Q&A]
Hello,

I have 2 questions related to the integrator cvodes.

1. Generally, cvodes integrates from t initial to t final. Is it possible to terminate the solver using some condition? For example, when one of my states reaches some value. This is more or less like the event command in MATLAB.

2. How to obtain the solution at different times? i.e., not only initial and final times. For example, if I would like to get the solutions at 0:0.1:tfinal --> every 0.1 second.

Thank you

**Comment by brunoml5** [comment]
Hi!

For your second question, which is easier, use this syntax in v.3.6.0:

```
t_grid = np.linspace(0.1, tfinal, tfinal/0.1)
F_horizon = ca.integrator('F', 'cvodes', dae, 0, t_grid, opts)
```

For your first question, I got creative and implemented events using the `if_else` command from Casadi. The idea is to add `t` as a state and then to integrate from 0 to 1 in `t2`. Then, I used the `if_else` command to "shutdown" the derivative when the event occurs.

I hope it helps!

```
import casadi as ca
import numpy as np

x = ca.MX.sym('x', 1)
u = ca.MX.sym('u', 1)
t = ca.MX.sym('t', 1)

Ts = 0.1
tfinal = 5

n_grid = int(tfinal/Ts)
t2_grid = np.linspace(1, n_grid, n_grid)

dx = u-x    # First order system

# Changing the ODE for t2
dx_dt2 = Ts*dx
dt_dt2 = Ts

dae = {'x':ca.vertcat(x, t), 'u':u, 'ode':ca.vertcat(dx_dt2, dt_dt2)}
F = ca.integrator('F', 'cvodes', dae, 0, t2_grid)

result = F(x0=[0, 0], u=1)

plot_x = result['xf'][0,:].full()
plot_t = result['xf'][1,:].full()

print(plot_t)    # You can see that the 't' state starts at 0 and goes to tfinal

event = ca.if_else(x>0.8, 0, 1)    # Let's create an event, if x>0.8, the result is zero

# I will multiply the dx with event so that if it happens, event=0 and dx will be "shutdown'
dx2_event = event*dx_dt2

dae = {'x':ca.vertcat(x, t), 'u':u, 'ode':ca.vertcat(dx2_event, dt_dt2)}
F_event = ca.integrator('F_event', 'cvodes', dae, 0, t2_grid)
result = F_event(x0=[0, 0], u=1)

plot_x_event = result['xf'][0,:].full()
plot_t_event = result['xf'][1,:].full()

print(plot_t_event)    # You can see that the 't' state is the same

import matplotlib.pyplot as plt
fig = plt.figure(dpi=150)

plt.plot(plot_t.T, plot_x.T, '.-', label='x')
plt.plot(plot_t_event.T, plot_x_event.T, 'x-', label='x_event')
plt.grid(True)
plt.xlabel('Time')
plt.ylabel('x')
plt.legend()
```

**Reply by jaeandersson** [reply]
This is a good answer. There is currently no event support for the CasADi integrators. @brunoml5's workaround makes sense.
